### PR TITLE
Core/DB/Quest: Battle for the Undercity

### LINKED ISSUE #204: [Quest] The Battle for Undercity [$15]
**Description**: Quest completely missing since there are no sniffs because the quest has been removed from retail.

**Current behaviour**: Simply, the script is missing

**Expected behaviour**: There are actually 2 quests, one for Alliance and one for Horde players.
Since describing the quests would take too long, I suggest to see this videos (one Alliance, one Horde)
https://www.youtube.com/watch?v=thvM9fd-pJA
https://www.youtube.com/watch?v=PxvQKCJagoI

**AC hash/commit**:  
3f25872719f7d81c3b8387a5e371a7944280af

**Operating system**:  Any

This quest is unfixable on Trinity Core, because there are no sniffs; we can start working on it together here anyway since this core accepts no-sniffs fixes.
## 

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/38728834-quest-the-battle-for-undercity?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Nefertumm**
##### CHANGES PROPOSED:

-  Imported the whole script of the quest from Dalaran-Wow.

###### ISSUES ADDRESSED:

Closes #204 

##### TESTS PERFORMED:

Compiled + tested on both sides, working.

##### HOW TO TEST THE CHANGES:

Alliance: [link](https://www.wowhead.com/quest=13377/the-battle-for-the-undercity#comments:id=863205)
Horde: [link](https://www.wowhead.com/quest=13267/the-battle-for-the-undercity#comments:id=387062)
Complete the previous quest and test. It's possible that the previous quests of the chain do not work properly. If not, open an issue.

##### Target branch(es):

Master

### CODE REVIEW

**Comment by BarbzYHOOL** [comment]
Awesome. Note that there are more things on this core to take from (notably SQL)

**Comment by comix1988** [comment]
premise: I confirm that there are some previous quests that do not work (will be reported)

Alliance side:
we have some small bugs that can be easily arranged:

1- fix Jaina (look at minutes 2:45 and 8:40), Varian (look at minutes 2:54) and stormwind elite (look at minutes 5:30) waypoints;
2- the boss grandapothecary has a base of attack too big, you can hit at a distance (look at minutes 10:10);
3- at the end of the mission, Jaina must immobilize everyone with an ice spell before teleporting them to stormwind
ghiaccio

video Link--->   https://youtu.be/vbbzoXATe7c

Horde side coming soon

**Comment by Nefertumm** [comment]
About waypoints: Some of them fails by vmaps i think. But if you see some of them with wrong Z position, feel free to let me know.
I will see the other things.

**Comment by comix1988** [comment]
I tested the quest also horde side and it works properly.
We need to fix the Sylvanas waypoints but I think it's the same problem as Jaina.

**Comment by Naxxramas-Server** [comment]
When i take the quest who say me to take the undercity portal , didn't see this portal , someone have the entry and the phase for this gameobject ? Probably removed it before ...

**Comment by comix1988** [comment]
the previous quest does not work, it will be reported

File: data/sql/updates/pending_db_world/rev_1550359453246256000.sql [hunk 1]
```diff
@@ -0,0 +1,1090 @@
...
+UPDATE `creature_template` SET InhabitType=4, minlevel=74, maxlevel=75 WHERE entry=@ALLIANCEPLANE;
+UPDATE `creature_template` SET minlevel=75, maxlevel=76 WHERE entry=@SWSOLDIER;
+
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@PLAGUEDFELBEAST;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@GUARDIAN;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@DREADLORD;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@BLIGHTDOCTOR;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@COLLABORATOR;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@BETRAYER;
+UPDATE `creature_template` SET unit_flags=0, minlevel=74, maxlevel=76 WHERE entry=@CHEMIST;
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1550359453246256000.sql]: please always prefer compact queries, e.g. `UPDATE whatever SET something WHERE entry IN (....)` whenever updates/inserts/deletions can be grouped

File: data/sql/updates/pending_db_world/rev_1550359453246256000.sql [hunk 2]
```diff
@@ -0,0 +1,1090 @@
...
+(@CGUID+54, 31576, 0, 0, 0, 1, 64, 0, 0, 1786.49, 251.731, 59.4535, 4.60986, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+55, 31576, 0, 0, 0, 1, 64, 0, 0, 1785.77, 244.768, 59.8426, 4.60986, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+56, 31576, 0, 0, 0, 1, 64, 0, 0, 1785.35, 240.702, 60.4073, 4.60986, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+57, 31576, 0, 0, 0, 1, 64, 0, 0, 1794.47, 246.206, 60.3204, 1.53974, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+58, 31576, 0, 0, 0, 1, 64, 0, 0, 1797.73, 254.984, 59.8279, 0.832876, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+59, 31576, 0, 0, 0, 1, 64, 0, 0, 1810.96, 257.323, 59.9768, 6.25369, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+60, 31576, 0, 0, 0, 1, 64, 0, 0, 1816.65, 258.938, 60.0562, 1.12976, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+61, 31576, 0, 0, 0, 1, 64, 0, 0, 1819.64, 265.263, 59.9507, 1.12976, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+62, 31576, 0, 0, 0, 1, 64, 0, 0, 1823.2, 269.927, 60.2139, 0.187279, 300, 0, 0, 5342, 0, 0, 0, 33554688, 0, 0),
+(@CGUID+63, 31576, 0, 0, 0, 1, 64, 0, 0, 1829.72, 270.662, 59.9834, 5.85314, 300, 0, 0, 5342, 0, 0, 0
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1550359453246256000.sql]: same as above, those inserts can be combined

Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1550359453246256000.sql]: same as above

Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1550359453246256000.sql]: same as above

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 3]
```diff
@@ -338,13 +343,3693 @@ class npc_parqual_fintallas : public CreatureScript
...
-    new npc_parqual_fintallas();
+public:
+    npc_varian_wrynn() : CreatureScript("npc_varian_wrynn") { }
+
+    bool OnGossipSelect(Player* player, Creature* creature, uint32 /*sender*/, uint32 action) override
+    {
+        player->PlayerTalkClass->ClearMenus();
+        switch (action)
+        {
+        case GOSSIP_ACTION_INFO_DEF + 1:
```
Comment (FrancescoBorzi) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: indentation

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 4]
```diff
@@ -338,13 +343,3693 @@ class npc_parqual_fintallas : public CreatureScript
...
+                    }
+                }
+            }
+        }
+
+        void JustSummoned(Creature* summoned) override
+        {
+            switch (summoned->GetEntry())
+            {
+            case NPC_GENERATOR:
```
Comment (FrancescoBorzi) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: indentation

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 5]
```diff
@@ -338,13 +343,3693 @@ class npc_parqual_fintallas : public CreatureScript
...
+            default:
+                break;
+            }
+        }
+
+        void SummonedCreatureDies(Creature* summon, Unit* /*killer*/) override
+        {
+            switch (summon->GetEntry())
+            {
+            case NPC_BLIGHTWORM:
```
Comment (FrancescoBorzi) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: indentation

Comment (FrancescoBorzi) [reply src/server/scripts/EasternKingdoms/zone_undercity.cpp]: and in other similar places too

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 6]
```diff
@@ -338,13 +343,3693 @@ class npc_parqual_fintallas : public CreatureScript
...
+                        bStepping = false;
+                        JumpToNextStep(0);
+                        break;
+                    case 54:
+                        Talk(WRYNN_SAY_APO_7);
+                        SetRun(false);
+                        JumpToNextStep(4 * IN_MILLISECONDS);
+                        break;
+                    case 55:
+                        if (Creature* putress = ObjectAccessor::GetCreature(*me, putressGUID))
```
Comment (FrancescoBorzi) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: ` fatal error: unused variable 'putress' [-Wunused-variable]`

**Comment by comix1988** [comment]
Any news about my report?

**Comment by Nefertumm** [comment]
About waypoints: it's because vmaps, it's not related to this pr.
About Jaina spell: It's working well on horde side, but on the alliance side i have no clue why it doesn't freezing the units.
About Apothecary: I'll take a look at his model_info

**Comment by Nefertumm** [comment]
@comix1988 i've changed the putress thing. Re-test again when you can

**Comment by comix1988** [comment]
I tried the quest again for both sides but I have a problem:
now when I accept the previous quests The Killing Time (ally) and A Life Without Regret (horde) I can not find the npcs to teach them anymore.
maybe I'm wrong?
I need these quests to start "battle for undecity"
creature
ac

**Comment by FrancescoBorzi** [comment]
@Naxxramas-Server have you tested this PR? if yes, please let us know :)

**Comment by Naxxramas-Server** [comment]
Yes that's why i approved it ^^

**Comment by FrancescoBorzi** [comment]
@Naxxramas-Server ok thanks for testing it and letting us know. The reason I asked is because we usually use the approve for code review, and the testers leave a comment when they test :)

**Comment by BarbzYHOOL** [comment]
Yes, please next time leave comment @Naxxramas-Server and thank you :P

**Pull request merged**